src/execute_linter.cc

```diff
@@ -13,55 +13,157 @@
+LinterResult ParseNoLintSingleComment(absl::string_view line,
+                                      const absl::string_view& sql,
+                                      LinterOptions* options, int position) {
   LinterResult result;
-  result.Add(CheckLineLength(sql));
-  result.Add(CheckParserSucceeds(sql));
-  result.Add(CheckSemicolon(sql));
-  result.Add(CheckUppercaseKeywords(sql));
-  result.Add(CheckCommentType(sql));
-  result.Add(CheckAliasKeyword(sql));
-  result.Add(CheckTabCharactersUniform(sql));
-  result.Add(CheckNoTabsBesidesIndentations(sql));
+  std::map<std::string, ErrorCode> error_map = GetErrorMap();
+
+  const LazyRE2 kRegex = {"\\s*(NOLINT|LINT)\\s*\\(([a-z ,-]*)\\)\\s*(.*)\\s*"};
+  std::string type = "";
+  std::string check_names = "";
+  std::string lint_comment = "";
+  bool matched =
+      RE2::FullMatch(line, *kRegex, &type, &check_names, &lint_comment);
+
+  // Found enabling or disabling type of comment.
+  if (matched) {
+    check_names.erase(
+        std::remove_if(check_names.begin(), check_names.end(), ::isspace),
+        check_names.end());
+
+    std::vector<std::string> names = absl::StrSplit(check_names, ',');
+    for (std::string check_name : names) {
+      // The name inside of parantheses is stored in 'check_name'
+      // If it is not valid add error, otherwise enable/disable position
+      if (!error_map.count(check_name)) {
+        result.Add(ErrorCode::kNoLint, sql, position,
+                   absl::StrCat("Unkown NOLINT error category: ", check_name));
+      } else {
+        const ErrorCode& code = error_map[check_name];
+        if (type == "NOLINT")
+          options->Disable(code, position);
+        else
+          options->Enable(code, position);
+      }
+    }
+  }
   return result;
 }
 
-}  // namespace
-}  // namespace zetasql::linter
+// This function will parse "NOLINT"(<CheckName>) syntax from a sql file.
+// If NOLINT usage errors occur, it counts as a lint
+// error and it will be returned in a result.
+LinterResult ParseNoLintComments(absl::string_view sql,
+                                 LinterOptions* options) {
+  LinterResult result;
+  for (int i = 0; i < static_cast<int>(sql.size()); ++i) {
+    if (sql[i] == '\'' || sql[i] == '"') {
+      // The sql is inside string. This will ignore sql part until
+      // the same type(' or ") of character will occur
+      // without \ in the beginning. For example 'a"b' is a valid string.
+      for (int j = i + 1; j < static_cast<int>(sql.size()); ++j) {
+        if (sql[j - 1] == '\\' && sql[j] == sql[i]) continue;
+        if (sql[j] == sql[i] || j + 1 == static_cast<int>(sql.size())) {
+          i = j;
+          break;
+        }
+      }
+      continue;
+    }
 
-int main(int argc, char* argv[]) {
-  if (argc < 1) {
-    std::cout << "Usage: execute_linter < <file_name>\n" << std::endl;
-    return 1;
-  }
+    // Ignore multiline comments.
+    if (i > 0 && sql[i - 1] == '/' && sql[i] == '*') {
```

Comment:
Please consider using regex here if it is possible.

Reply:
Again, We need to hold the location, There are 5 types: ((/* (.*) */) | ('(.*)') | ("(.*)") | (.*) | (//(.*)\n)) in which can contain each other, and groups are extremely important. To cover every case precisely, I trust this implementation more. Also, There are cases like line delimeter is different, then line comment regex (//(.*)\n) should be different. For any update like this, changing it in the code is more clear than it would in the regex.

Reply:
github changed the regex with italic characters :) it was: ``` ((/* (.*) */) | ('(.*)') | ("(.*)") | (.*) | (//(.*)\n)) ```

Reply:
I just checked simple case: ```/**/SELECT/**//**/``` and it is not working for regex: ```((\/\*(.*)\*\/)|SELECT)*```
There are 3 groups: ```/**/SELECT/**//**/```, ```/**/SELECT/**//**/``` and ```*/SELECT/**//*```.
Expected groups was (4 groups): ```/**/``` ```SELECT``` ```/**/``` ```/**/```
So it just gets everytinh to first group and doesn't parse sections, so probably we can't use it.